### ISSUE TITLE
How to further fix to purge_old_records task (?)

### ISSUE THREAD

**Issue by jrgm** [mozilla-services/tokenserver#170]
I was running the updated purge_old_records task in production, and the new changes were working fine, until it began hitting "rich veins" of downed nodes.

The task would pull up 1000 candidate rows and then immediately discard them all for downed nodes and re-select another result set. This would peg the unacceptably cpu on the database server.

I tried this hack to get around this problem:
```
diff --git a/tokenserver/scripts/purge_old_records.py b/tokenserver/scripts/purge_old_records.py
index dcb248f..78fcd35 100644
--- a/tokenserver/scripts/purge_old_records.py
+++ b/tokenserver/scripts/purge_old_records.py
@@ -79,6 +79,12 @@ def purge_old_records(config_file, grace_period=-1, max_per_loop=10,
                         delete_service_data(config, service, row,
                                             timeout=request_timeout)
                         backend.delete_user_record(service, row.uid)
+                    else:
+                        logger.info("Punting on this row: uid %s on %s (%d)",
+                                    row.uid, row.node, row.downed)
+                        # Essentially pause for 60 seconds if all the nodes
+                        # in these return rows are downed.
+                        time.sleep(60.0/len(rows))
                 if len(rows) < max_per_loop:
                     break
     except Exception:
```
... but that settled into a state where it never found any records to purge (but didn't peg the database cpu).

I tried increasing the limit and offset and occasionally it found pockets to purge, but really wasn't making a lot of progress.

I think this may need a change in the _GET_OLD_USER_RECORDS_FOR_SERVICE query. Something like:
```
diff --git a/tokenserver/assignment/sqlnode/sql.py b/tokenserver/assignment/sqlnode/sql.py
index b602f06..9c7166f 100644
--- a/tokenserver/assignment/sqlnode/sql.py
+++ b/tokenserver/assignment/sqlnode/sql.py
@@ -112,15 +112,19 @@ where
 
 _GET_OLD_USER_RECORDS_FOR_SERVICE = sqltext("""\
 select
     uid, email, generation, keys_changed_at, client_state,
     nodes.node, nodes.downed, created_at, replaced_at
 from
-    users left outer join nodes on users.nodeid = nodes.id
+    users
+use index (replaced_at)
+left outer join nodes on users.nodeid = nodes.id
 where
     users.service = :service
+and
+    nodes.downed != 1
 and
     replaced_at is not null and replaced_at < :timestamp
 order by
     replaced_at desc, uid desc
 limit
     :limit
 offset
     :offset
```
That is, filter out nodes.downed in the query. It also needed some help on the index hint to not choose an expensive plan.

I dunno. rfk: Thoughts on how to deal with this?

**Comment by jrgm** [comment]
Maybe option #1, the sleep, and just going into a hover pattern is good enough (combined with increased limit/offset to search deeper). But doesn't really feel right.

**Comment by rfk** [comment]
Are these downed nodes ever going to come back, or are they permanently decommissioned? We want to make sure we don't abandon them permanently in a state where we'll never try to delete the data, but will leave the data hanging around on a node somewhere.  If the node is never coming back, we should delete the old records in the db.

I think what you propose here seems fine, don't select them in the first place if we're going to ignore them. We'd have to double-check for other callers of this method to ensure they won't be impacted, but I think it's mostly tests and we can adjust as needed.

**Comment by rfk** [comment]
IIUC the `sleep` option might result in just selecting the same set of downed rows over and over again, until some new rows get marked as replaced and become available for cleanup. This in turn would risk ignoring older replaced rows that happen to be "hidden behind" the downed ones.

**Comment by jrgm** [comment]
> Are these downed nodes ever going to come back, or are they permanently decommissioned?

I don't know right now, but was going to check tomorrow with Erik. It was all the same node.

**Comment by jrgm** [comment]
> We'd have to double-check for other callers of this method to ensure they won't be impacted, but I think it's mostly tests and we can adjust as needed.

Yeah, I haven't actually written this as a patch and I'll probably have to adjust tests. Question: are there "external" users of this module?

**Comment by jrgm** [comment]
> IIUC the sleep option might result in just selecting the same set of downed rows over and over again, until some new rows get marked as replaced and become available for cleanup. This in turn would risk ignoring older replaced rows that happen to be "hidden behind" the downed ones.

Yeah, good point. It could get in a state where it was always working (of punting) on the head of the result set, and never reaching into the backlog.

**Comment by rfk** [comment]
> Question: are there "external" users of this module?

Note that I know of.

**Comment by jrgm** [comment]
Hrm. How does one use `with_hint` in our SQLAlchemy sql.py code? Because I can't just make the `use index` change I suggested above (tests fail on that syntax).

**Comment by jrgm** [comment]
So, I tried this as a simple test of using mysql hints:
```
from sqlalchemy import create_engine, MetaData, Table, Column, Integer, String, BigInteger, Index
from sqlalchemy.sql import select, text as sqltext

engine = create_engine('mysql://root@localhost/test')
conn = engine.connect()
meta = MetaData()

users = Table(
  'users',
  meta,
  Column('uid', BigInteger, primary_key = True),
  Column('service', Integer),
  Column('email', String(255)),
  Column('replaced_at', BigInteger),
)
Index('replaced_at_idx', users.c.service, users.c.replaced_at)
meta.create_all(engine)

query = sqltext('uid, service, replaced_at from users')
stmt = select([query])
stmt = stmt.with_statement_hint('use index (replaced_at_idx)', dialect_name='mysql')
stmt = stmt.where(sqltext('service = 1'))

res = conn.execute(stmt).fetchall() 
for row in res:
  print row
```
and I get:
```
sqlalchemy.exc.ProgrammingError: (_mysql_exceptions.ProgrammingError) (1064, "You have an 
error in your SQL syntax; check the manual that corresponds to your MySQL server version for 
the right syntax to use near 'use index (replaced_at_idx)' at line 2")
[SQL: SELECT uid, service, replaced_at from users 
WHERE service = 1 use index (replaced_at_idx)]
```

Notice that it simply appends the `use index` to the end of the query. But I don't grok the documentation for how to use this (and possibly this feature just doesn't work since the sqlalchemy tests do not test for a select with a where clause).

**Comment by rfk** [comment]
Using [`with_hint`](https://docs.sqlalchemy.org/en/13/orm/query.html#sqlalchemy.orm.query.Query.with_hint) seems to do the job for me, e.g.:

```
select([users.c.uid, users.c.service, users.c.replaced_at]).with_hint(users, "USE INDEX (replaced_at_idx)", dialect_name="mysql")).where(users.c.service == 1)
```

IIUC you have to use this variant which accepts the `users` table as a additional argument, so that it knows to attach the hint to the appropriate table in the select clause.

**Comment by jrgm** [comment]
Thanks. Hmm, I had tried `with_hint`, but not with the `users.c.*` syntax. I'll work with that to create a patch.
